### ISSUE TITLE
slowness of test/core/skip-stack-guard-page.wast

### ISSUE THREAD

**Issue by titzer** [WebAssembly/function-references#78]
This spec test has gotten a lot slower with the function references and gc proposals, taking over 30 seconds on my machine. The rest of the tests are in the millisecond range. I looked into why this is and if there was something that could be done to make the test faster. E.g. I suspected I could nerf this test to run in less time, but alas, it was not so easy because it tries to test different nesting depths and ultimate needs a lot of (live) locals to function. I concluded that it was a change in the way locals are implemented (I suspect due to `let`?) in the reference interpreter.

I was wondering if there's something I hadn't thought of to make this test faster, or if there is an inherent interpreter speed limit here.

Thoughts?

**Comment by rossberg** [comment]
The problem actually was validation. Following the paper rules, the interpreter updates the local environment after every instruction to apply the new init status. However, due to the way the typing rule for instruction sequences was formulated as a left fold, it redid that for all previous variables after every instruction, which let to quadratic behaviour.

Fixed by turning the rule into a right fold. Time for the test went down to 2s on my machine (vs 23s before), and the whole test suite is likewise faster. :)

@conrad-watt, I switched around the rule for instruction sequences in the spec accordingly, see [here](https://github.com/WebAssembly/function-references/commit/d07e0decdb709f1cfe431c72d5df63b404eb13af). IIRC, somebody in Cambridge already asked me why it was the other way, don't remember if that was you.
